llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -3009,6 +3037,22 @@ SDValue RISCVTargetLowering::LowerOperation(SDValue Op,
     unsigned NewOp = ToCap ? ISD::INTTOPTR : ISD::PTRTOINT;
     return DAG.getNode(NewOp, DL, Op.getValueType(), Op0);
   }
+  case ISD::PTRTOINT: {
+    SDValue Op0 = Op.getOperand(0);
+    if (Op0.getValueType().isFatPointer() &&
+        !Subtarget.hasCheriISAv8Semantics() &&
+        !RISCVABI::isCheriPureCapABI(Subtarget.getTargetABI())) {
```

Comment:
Why not guard setOperandAction with this?

Reply:
Done, and kept an assertion just in case.